### ISSUE TITLE
Plot arguments aren't always working

### ISSUE THREAD

**Issue by Nylio-prog** [ultralytics/ultralytics#5006]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Predict

### Bug

When plotting the result of an inference, some parameters seem to not be applied. 
`        result = model(frame, agnostic_nms=True)[0]

        annotated_frame = result.plot(
            pil=True, line_width=4, font_size=30)

        cv2.imshow("YOLOv8 Inference", annotated_frame)`
This is the piece of code  I was trying to run and when trying to tweak some paramaters such as font_size and font, it doesn't change anything and still remains the same. I'm sure that the font_size is passed until the annotator but I don't understand why it doesn't work. 
![image](https://github.com/ultralytics/ultralytics/assets/77471647/a1f7d913-c005-4e49-885e-85ab8e7264d8)
Whatever values I pass in, I still get the same font_size. However, the line_width or pil works well.

### Environment

Ultralytics YOLOv8.0.180  Python-3.11.5 torch-2.0.1+cu118 CUDA:0 (NVIDIA GeForce GTX 1080, 8192MiB)
Setup complete  (8 CPUs, 15.9 GB RAM, 445.6/464.7 GB disk)

### Minimal Reproducible Example

`        result = model(frame, agnostic_nms=True)[0]

        annotated_frame = result.plot(
            pil=True, line_width=4, font_size=30)

        cv2.imshow("YOLOv8 Inference", annotated_frame)`

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@Nylio-prog I'm able to change font size and line width with pil=True, i.e.:

![test](https://github.com/ultralytics/ultralytics/assets/26833433/9923ff4c-7b36-4cf0-bde8-09a8766bfd64)

**Comment by glenn-jocher** [comment]
For cv2 the font size is not independent of the font thickness, so we've connected the two attributes together, by increase linewidth you'll also increase fontsize.

**Comment by Nylio-prog** [comment]
Thank you for confirming me that it works but on my end it still doesn't work. There must be something wrong in my code then, this is my whole script and I'm going to investigate on it : 
`import cv2
import argparse
import torch
import time

from ultralytics import YOLO


def main():
    """
    Annotates a video with object detection bounding boxes and displays real-time FPS.

    This script processes a video using YOLO object detection and annotates it with bounding boxes
    around detected objects. It also displays the real-time frames per second (FPS) of the video.

    Args:
        -i (str): Path to the input video file.
        -o (str): Path to the output video file.
        --not-show (bool, optional): Show the annotation in real-time (default: False).
        -fps (float, optional): Desired frames per second (FPS) for the output video (default: 25.0).
    """

    # Parse command line arguments
    parser = argparse.ArgumentParser()
    parser.add_argument("-i", type=str, default="input_files/video.mp4",
                        help="Path to the input video file")
    parser.add_argument("-o", type=str, default="video_annotated.mp4",
                        help="Output video file name, will always be in output_files folder")
    parser.add_argument("-m", type=str, default="best.pt",
                        help="Model to use")
    parser.add_argument("--not-show", action="store_true",
                        help="Show the annotation in not")
    parser.add_argument("-fps", type=float, default=25.0,
                        help="Desired fps for the output video")
    args = parser.parse_args()

    global_start_time = time.time()

    # Check if CUDA (GPU support) is available
    use_gpu = torch.cuda.is_available()

    if use_gpu:
        device_name = torch.cuda.get_device_name(0)
        print(f"GPU Device: {device_name}")
    else:
        print("No GPU detected. Using CPU.")

    # Always putting the file in the output folder
    args.o = "output_files/" + args.o

    # Create VideoCapture objects for input and output videos
    cap = cv2.VideoCapture(args.i)
    fourcc = cv2.VideoWriter_fourcc(*"mp4v")
    out = cv2.VideoWriter(args.o, fourcc, args.fps, (int(
        cap.get(3)), int(cap.get(4))))  # cap.get(3) returns width

    model = YOLO(args.m)

    prev_end_time = 0
    start_time = 0

    while True:
        start_time = time.time()

        ret, frame = cap.read()

        if not ret:
            break

        result = model(frame, agnostic_nms=True)[0]

        # Visualize the results on the frame
        annotated_frame = result.plot(
            pil=True, line_width=4, font_size=40)

        # Can't compute it for first frame
        if (prev_end_time > 0 and not (args.not_show)):
            # Calculate the FPS of frame - 1
            fps = 1.0 / elapsed_time

            # Add FPS text to the top-left corner of the frame
            fps_text = f"FPS: {fps:.2f}"
            # Add FPS text to the top-left corner of the frame
            cv2.putText(annotated_frame, fps_text, (10, 30),
                        cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 0, 0), 2, cv2.LINE_AA)

        if (not (args.not_show)):
            # Display the annotated frame
            cv2.imshow("YOLOv8 Inference", annotated_frame)

        if cv2.waitKey(30) == 27:
            break

        # Write the frame with bounding boxes to the output video
        out.write(annotated_frame)

        prev_end_time = time.time()
        elapsed_time = prev_end_time - start_time

    cap.release()
    out.release()
    cv2.destroyAllWindows()

    global_elapsed_time = time.time() - global_start_time
    print(f"The whole process took {global_elapsed_time} seconds to execute")


if __name__ == "__main__":
    main()
`

**Comment by RizwanMunawar** [comment]
@Nylio-prog Thank you for providing the code. I have used the same code for testing, and I can observe the changes in both the bounding box size and the font size. The code adjustments for `font_size` and `line_width` and the resulting output are detailed below.

### Code
```python
import cv2
import argparse
import torch
import time

from ultralytics import YOLO


def main():
    # Parse command line arguments
    parser = argparse.ArgumentParser()
    parser.add_argument("-i", type=str, default="input_files/video.mp4",
                        help="Path to the input video file")
    parser.add_argument("-o", type=str, default="video_annotated.mp4",
                        help="Output video file name, will always be \
                            in output_files folder")
    parser.add_argument("-m", type=str, default="best.pt",
                        help="Model to use")
    parser.add_argument("--not-show", action="store_true",
                        help="Show the annotation in not")
    parser.add_argument("-fps", type=float, default=25.0,
                        help="Desired fps for the output video")
    args = parser.parse_args()

    global_start_time = time.time()
    elapsed_time = 0

    # Check if CUDA (GPU support) is available
    use_gpu = torch.cuda.is_available()

    if use_gpu:
        device_name = torch.cuda.get_device_name(0)
        print(f"GPU Device: {device_name}")
    else:
        print("No GPU detected. Using CPU.")

    # Always putting the file in the output folder
    args.o = "output_files/" + args.o

    # Create VideoCapture objects for input and output videos
    cap = cv2.VideoCapture(args.i)
    fourcc = cv2.VideoWriter_fourcc(*"mp4v")
    out = cv2.VideoWriter(args.o, fourcc, args.fps, (int(
        cap.get(3)), int(cap.get(4))))  # cap.get(3) returns width

    model = YOLO(args.m)

    prev_end_time = 0
    start_time = 0

    while True:
        start_time = time.time()

        ret, frame = cap.read()

        if not ret:
            break

        result = model(frame, agnostic_nms=True)[0]

        # Visualize the results on the frame
        annotated_frame = result.plot(
            pil=True, line_width=1, font_size=5)

        # Can't compute it for first frame
        if (prev_end_time > 0 and not (args.not_show)):
            # Calculate the FPS of frame - 1
            fps = 1.0 / elapsed_time

            # Add FPS text to the top-left corner of the frame
            fps_text = f"FPS: {fps:.2f}"
            # Add FPS text to the top-left corner of the frame
            cv2.putText(annotated_frame, fps_text, (10, 30),
                        cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 0, 0), 2, cv2.LINE_AA)

        if (not (args.not_show)):
            # Display the annotated frame
            cv2.imshow("YOLOv8 Inference", annotated_frame)

        if cv2.waitKey(30) == 27:
            break

        # Write the frame with bounding boxes to the output video
        out.write(annotated_frame)

        prev_end_time = time.time()
        elapsed_time = prev_end_time - start_time

    cap.release()
    out.release()
    cv2.destroyAllWindows()

    global_elapsed_time = time.time() - global_start_time
    print(f"The whole process took {global_elapsed_time} seconds to execute")


if __name__ == '__main__':
    main()
```


### Output: (line_width: 1, font_size=5)

![Screenshot 2023-09-21 130047](https://github.com/ultralytics/ultralytics/assets/62513924/b75520a4-e436-462f-8c01-edb97ad339df)


### Output: (line_width: 5, font_size=40)

![Screenshot 2023-09-21 130124](https://github.com/ultralytics/ultralytics/assets/62513924/540274ee-b3f8-4dc1-bf32-5ba1051f92e7)


Thanks.
Ultralytics Team!

**Comment by RizwanMunawar** [comment]
I am closing this issue. Please reopen if the issue persists

**Comment by drbilal216** [comment]
Nicely explained, i have a problem related to this, i want to add margin to (x,y) in the above code so that person text can be plot in the desired position. Im working on a PPE project and all the classes are displayed really close, want to adjust those as for my need. I have manually done this, but wondering how to do this using result.plot method.

**Comment by glenn-jocher** [comment]
@drbilal216 the `result.plot()` method utilizes default settings for plotting bounding box labels in a way that closely aligns with the top-left corner of the detected object's bounding boxes, and currently, it might not provide an option for custom padding or margins directly.

However, if you find that the labels are too close together, you might consider modifying the source code of the method where the labels are actually drawn onto the image. You can manually adjust the label positioning by altering the x and y coordinates used to draw the text. Remember that any changes to the source code will affect how all labels are drawn, not just for specific classes.

Alternatively, if you prefer using the existing functionality without modifying the source code, create a derived class that extends the annotation functionality of the `result.plot()` method and override the function responsible for drawing text to incorporate custom margins.

For future updates, considerations for improved customization of the annotations could be taken into account. It could be helpful to file a feature request in the GitHub repository issue tracker outlining your specific needs for adjusting label margins. The Ultralytics team and contributing community often take user feedback into account when planning new features.

As a side note, if you have already figured out a manual way to add margins to the (x,y) positions in your annotations, you could consider sharing this with the community by opening a pull request or creating a fork of the repository with your custom annotations. It’s always great to see how users adapt tools for specific use cases, and sharing your solution might help others with similar needs!

In the meantime, for detailed annotation control, it may be necessary to work directly with the underlying drawing libraries that `result.plot()` utilizes, which are typically either PIL (Pillow) or OpenCV, depending on whether you use the `pil` flag.

Remember to thoroughly test any changes to ensure they perform as expected across your datasets.

**Comment by drbilal216** [comment]
@glenn-jocher, it would be nice if these are added, set custom confidence for each class, set custom color for bounding boxes, add margin to display text at desired place and if possible in boot or hand class, display text of right hand or boot at right side and display the text of left boot or hand on left side, I have wrote custom code to do all of this, I’m working on a project for someone so i can’t share much of the code,  I’m sharing the margin part of the code.
```python
import cv2
from ultralytics import YOLO

model = YOLO("best.pt")
cap = cv2.VideoCapture("video_path.mp4")

# Define custom offsets for each class
class_custom_offsets = {
    'helmet': (0+50, 0-10),
    'glasses': (0+60, 0),
    '...': (0-180, 0),
    '...': (0, 0),
    '...': (0, 0),
    '...': (0-205, 0-5)
}
class_names = ['helmet','glasses','...','...','...','...']

while cap.isOpened():
    ret, frame = cap.read()
    if ret:
        results = model(frame, conf=0.3,verbose=False)
        annotated_frame = results[0].plot()     
        for result in results:
            boxes = result.boxes
            cls = boxes.cls.tolist()
            xyxy = boxes.xyxy
            conf = boxes.conf
            for i, class_index in enumerate(cls):
                class_name = class_names[int(class_index)]
                confidence = conf[i]
                x, y, w, h = map(int, xyxy[i])
                cv2.rectangle(frame, (x, y), (w, h), (0, 0, 255), 2)
                text_offset_x, text_offset_y = class_custom_offsets.get(class_name, (0, 0))
                org = (x + text_offset_x, y + text_offset_y)
                label = f"{class_name}: {confidence:.2f}"
                cv2.putText(frame, label, org, cv2.FONT_HERSHEY_SIMPLEX, 1.0, (0, 255, 0), thickness=3, lineType=cv2.LINE_AA)
        cv2.imshow("Video", frame)
        if cv2.waitKey(1) & 0xFF == ord("q"):
            break
cap.release()
cv2.destroyAllWindows()
```

**Comment by glenn-jocher** [comment]
@drbilal216 Thank you for sharing the margin adjustment part of your code. It's great to see how you've personalized object detection to suit your project’s specific needs. It indeed would be valuable for users to have more control over the annotation process directly through the `result.plot()` method, allowing for easier customization of the detection output.

I understand your suggestion for more granular control over the visual output, such as setting custom confidence per class, custom color for bounding boxes, adding margins for text display, and specific placement of text labels for certain classes like 'left hand' or 'right boot'. These are excellent feature ideas that would enhance user experience and tailor the display to complex scenarios, particularly in sectors such as industrial safety for PPE detection.

While these features are not currently part of the standard `result.plot()` method, I recommend opening a feature request on the Ultralytics YOLOv8 GitHub repository. The Ultralytics team is always looking for ways to improve the model and make it more user-friendly, and direct feedback from users like you is invaluable for guiding future developments.

Additionally, your approach by writing custom code shows a proactive step. Although you are working on a project with confidential constraints, your contributions, like the code snippet shared, are appreciated as they ignite discussions that could lead to the implementation of such features in future versions.

I'm glad to hear that you found a solution that works for your project, and I want to encourage you to stay engaged with the Ultralytics community. The team considers all suggestions, and your input could help shape future updates! Your efforts to enhance the usability of the YOLO models for specialized tasks further demonstrate the adaptability and flexibility of these sophisticated AI tools. Keep it up! 🚀✨
